pkg/dyninst/irgen/irgen.go

```diff
@@ -1794,23 +1832,25 @@ func populateEventExpressions(
 		default:
 			panic(fmt.Sprintf("unexpected event kind: %v", event.Kind))
 		}
-		variableSize := variable.Type.GetByteSize()
-		expr := &ir.RootExpression{
-			Name:   variable.Name,
-			Offset: uint32(0),
-			Kind:   variableKind,
-			Expression: ir.Expression{
-				Type: variable.Type,
-				Operations: []ir.ExpressionOp{
-					&ir.LocationOp{
-						Variable: variable,
-						Offset:   0,
-						ByteSize: uint32(variableSize),
-					},
-				},
-			},
-		}
+		expr := newRootExpressionForVariable(variable, variableKind)
 		expressions = append(expressions, expr)
+
+		// Check if this variable is relevant to a template segment.
+		if segmentIndex, ok := variableNamesUsedInTemplates[variable.Name]; ok {
+			seg := probe.Template.Segments[segmentIndex]
+			jsonSeg, ok := seg.(ir.JSONSegment)
+			if !ok {
+				// This should not happen.
+				continue
+			}
```

Comment:
I can avoid this cast by adding some complexity to the TemplateSegment, but I think this is fine. 

The code to avoid this from happening is in collectAllSegmentVariables